ibis/backends/base/sql/alchemy/registry.py

```diff
@@ -300,69 +299,73 @@ def _endswith(t, op):
+    return an.windowize_function(new_expr, frame=new_frame)
 
 
-def _window(t, op):
-    arg, window = op.args
-    reduction = t.translate(arg)
+def _translate_window_boundary(boundary):
+    if boundary is None:
+        return None
 
-    window_op = arg
+    if isinstance(boundary.value, ops.Literal):
+        if boundary.preceding:
+            return -boundary.value.value
+        else:
+            return boundary.value.value
 
-    if isinstance(window_op, ops.CumulativeOp):
-        arg = _cumulative_to_window(t, arg, window).op()
-        return t.translate(arg)
+    raise com.TranslationError("Window boundaries must be literal values")
 
-    if window.max_lookback is not None:
-        raise NotImplementedError(
-            'Rows with max lookback is not implemented '
-            'for SQLAlchemy-based backends.'
-        )
 
-    # Checks for invalid user input e.g. passing in tuple for preceding and
-    # non-None value for following are caught and raised in expr/window.py
-    # if we're here, then the input is valid, we just need to interpret it
-    # correctly
-    if isinstance(window.preceding, tuple):
-        start, end = (-1 * x if x is not None else None for x in window.preceding)
-    elif isinstance(window.following, tuple):
-        start, end = window.following
-    else:
-        start = -window.preceding if window.preceding is not None else window.preceding
-        end = window.following
+def _window_function(t, window):
+    if isinstance(window.func, ops.CumulativeOp):
+        func = _cumulative_to_window(t, window.func, window.frame).op()
+        return t.translate(func)
+
+    reduction = t.translate(window.func)
 
     # Some analytic functions need to have the expression of interest in
     # the ORDER BY part of the window clause
-    if isinstance(window_op, t._require_order_by) and not window._order_by:
-        order_by = t.translate(window_op.args[0])
+    if isinstance(window.func, t._require_order_by) and not window.frame.order_by:
+        order_by = t.translate(window.func.arg)  # .args[0])
     else:
-        order_by = [t.translate(arg) for arg in window._order_by]
+        order_by = [t.translate(arg) for arg in window.frame.order_by]
 
-    partition_by = [t.translate(arg) for arg in window._group_by]
+    partition_by = [t.translate(arg) for arg in window.frame.group_by]
+
+    if isinstance(window.frame, ops.RowsWindowFrame):
+        if window.frame.max_lookback is not None:
+            raise NotImplementedError(
+                'Rows with max lookback is not implemented for SQLAlchemy-based '
+                'backends.'
+            )
+        how = 'rows'
+    elif isinstance(window.frame, ops.RangeWindowFrame):
+        how = 'range_'
+    else:
+        raise NotImplementedError(type(window.frame))
+
+    if t._forbids_frame_clause and isinstance(window.func, t._forbids_frame_clause):
+        # some functions don't support frame clauses
```

Comment:
It's more accurate to say "some functions on some backends". No need to change anything, just want to capture that it's both a limitation/design choice by the backend as well as the particular function.

Reply:
Updated the comment.